learning/tour-of-beam/frontend/lib/auth/notifier.dart

```diff
@@ -44,33 +40,31 @@ class AuthNotifier extends ChangeNotifier {
     unawaited(_read());
   }
 
+  String? get token => _token;
+
   AuthStage get authStage => _authStage;
 
-  Future<void> signIn(AuthMethod authMethod) async {
-    if (_authStage == AuthStage.unauthenticated) {
-      // TODO(nausharipov): is switch better here than _authProviders?
-      final UserCredential userCredential = await FirebaseAuth.instance
-          .signInWithPopup(_authProviders[authMethod]!);
-      final User? user = userCredential.user;
+  bool get isAuthenticated => authStage == AuthStage.authenticated;
 
-      if (user != null) {
-        await _updateStorageToken(user.uid);
-      }
-      notifyListeners();
+  Future<void> signIn(AuthMethod authMethod) async {
+    if (_authStage == AuthStage.authenticated) {
+      return;
     }
+    await FirebaseAuth.instance.signInWithPopup(_authProviders.get(authMethod));
+    await _updateStorageToken();
   }
 
   Future<void> signOut() async {
     await FirebaseAuth.instance.signOut();
-    await _updateStorageToken(null);
+    await _updateStorageToken();
     notifyListeners();
   }
 
-  Future<void> _updateStorageToken(String? value) async {
-    // TODO(nausharipov): use FirebaseAuth.instance.currentUser instead?
+  Future<void> _updateStorageToken() async {
+    final token = await FirebaseAuth.instance.currentUser?.getIdToken();
     await _storage.write(
       key: _tokenStorageKey,
-      value: value,
+      value: token,
     );
     await _read();
```

Comment:
If `_read()` is split to `_read() + _verifyToken(String token)`, this way there's no need to read here.